opacus/tests/grad_samples/embedding_norm_sample_test.py

```diff
@@ -0,0 +1,218 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import unittest
+
+from opacus.grad_sample import embedding_norm_sample
+import torch
+import torch.nn as nn
+
+
+class TestComputeEmbeddingNormSample(unittest.TestCase):
+
+    def test_compute_embedding_norm_sample(self):
+        # Define the embedding layer
+        embedding_dim = 1
+        vocab_size = 3
+        embedding_layer = nn.Embedding(vocab_size, embedding_dim)
+
+        # Manually set weights for the embedding layer for testing
+        embedding_layer.weight = nn.Parameter(
+            torch.tensor([[0.1], [0.2], [0.3]], dtype=torch.float32)
+        )
+
+        # Example input ids (activations). Shape: [3, 2]
+        input_ids = torch.tensor([[1, 1], [2, 0], [2, 0]], dtype=torch.long)
+
+        # Example gradients with respect to the embedding output (backprops).
+        # Shape: [6, 1]
+        grad_values = torch.tensor(
+            [[0.2], [0.2], [0.3], [0.1], [0.3], [0.1]], dtype=torch.float32
+        )
+
+        # Simulate backprop through embedding layer
+        backprops = grad_values
+
+        # Wrap input_ids in a list as expected by the norm sample function
+        activations = [input_ids]
+
+        # Call the function under test
+        result = embedding_norm_sample.compute_embedding_norm_sample(
+            embedding_layer, activations, backprops
+        )
+
+        # Expected norms
+        expected_norms = torch.tensor([0.4000, 0.3162, 0.3162], dtype=torch.float32)
+
+        # Extract the result for the embedding layer weight parameter
+        computed_norms = result[embedding_layer.weight]
+
+        # Verify the computed norms match the expected norms
+        torch.testing.assert_close(computed_norms, expected_norms, atol=1e-4, rtol=1e-4)
+
+    def test_compute_embedding_norm_sample_with_non_one_embedding_dim(self):
+        # Define the embedding layer
+        embedding_dim = 2
+        vocab_size = 3
+        embedding_layer = nn.Embedding(vocab_size, embedding_dim)
+
+        # Manually set weights for the embedding layer for testing
+        embedding_layer.weight = nn.Parameter(
+            torch.tensor([[0.1], [0.2], [0.3]], dtype=torch.float32)
+        )
+
+        # Example input ids (activations). Shape: [6, 1, 1].
+        input_ids = torch.tensor(
+            [[[1]], [[1]], [[2]], [[0]], [[2]], [[0]]], dtype=torch.long
+        )
+
+        # Example gradients per input id, with embedding_dim=2.
+        # Shape: [6, 1, 1, 2]
+        grad_values = torch.tensor(
```

Comment:
Similar here, maybe we should distinguish between grad_values and backprops, since they have different shapes?